llvm/lib/Target/AIE/AIELiveRegs.h

```diff
@@ -0,0 +1,62 @@
+//===- AIELiveRegs.h - Liveness Analysis logic -*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Class providing services for Liveness Analysis.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_LIB_TARGET_AIE_AIELIVEREGS_H
+#define LLVM_LIB_TARGET_AIE_AIELIVEREGS_H
+
+#include "llvm/CodeGen/LivePhysRegs.h"
+#include "llvm/CodeGen/MachineBasicBlock.h"
+#include "llvm/CodeGen/MachineFunction.h"
+#include "llvm/CodeGen/TargetRegisterInfo.h"
+#include <map>
+#include <queue>
+#include <set>
+
+namespace llvm::AIE {
+
+class LiveRegs {
+  // Queue to manage the order of MBBs to be processed.
+  std::queue<const MachineBasicBlock *> WorkList;
+
+  // Set to ensure MBBs are added to the WorkList only once.
+  std::set<const MachineBasicBlock *> WorkListSet;
+
+  // Pointer to Target Register Information, used for initializing livein
+  // registers.
+  const TargetRegisterInfo *TRI;
+
+  // Pops a Machine Basic Block from the work list, removes it from the work
+  // list set, and returns the popped MBB.
+  const MachineBasicBlock *popFromWorkList();
+
+  // Computes the live-in registers for a given Machine Basic Block.
+  void computeBlockLiveIns(const MachineBasicBlock *MBB,
+                           LivePhysRegs &CurrentLive);
+
+  // Checks if the CurrentLive is the same as the liveins of the given MBB.
+  // If they are different, it updates the liveins of the MBB and returns false.
+  bool liveRegsAreEqual(LivePhysRegs &CurrentLive,
+                        const MachineBasicBlock *MBB);
+
+  // Adds a Machine Basic Block to the work list if it is not already present.
+  void addToWorkList(const MachineBasicBlock *MBB);
+
+public:
+  // Mapping from Machine Basic Blocks to their livein registers.
+  std::map<const MachineBasicBlock *, LivePhysRegs> LiveIns;
+
+  LiveRegs(const MachineFunction *MF);
```

Comment:
Future: It may be a bit expensive, but you could call this lazily from the DAG mutator if you can't reach InterblockScheduling